### PR TITLE
Font edging and hinting moved in settings

### PR DESCRIPTION
**Pull request by Nawy**
This PR adds support to Font Edging and Font Hinting settings from vim configuration files.

Neovide renders with too bright and too fat fonts on macOS and probably other systems. These settings bring flexibility to adjust Neovide's font loading without rebuilding.

Problem:
https://github.com/neovide/neovide/issues/1325

## What kind of change does this PR introduce?
- Feature
- Documentation

## Did this PR introduce a breaking change? 
- No

### CODE REVIEW

File: src/renderer/mod.rs [hunk 1]
```diff
@@ -39,7 +39,7 @@ pub struct RendererSettings {
     floating_blur_amount_y: f32,
     debug_renderer: bool,
     profiler: bool,
-    underline_automatic_scaling: bool,
+    underline_automatic_scaling: bool
```
Comment (MultisampledNight) [inline src/renderer/mod.rs:42]: You might want to consider running `cargo fmt`, which will format the whole repo with the Rust community's defaults (which include a trailing comma). Or you set `g:rustfmt_autosave` to `v:true` directly.

File: src/window/settings.rs [hunk 2]
```diff
@@ -35,6 +37,8 @@ impl Default for WindowSettings {
             touch_drag_timeout: 0.17,
             background_color: "".to_string(),
             confirm_quit: true,
+            font_edging: "AntiAlias".to_string(),
+            font_hinting: "Full".to_string()
```
Comment (MultisampledNight) [inline src/window/settings.rs:41]: All other settings are in lowercase, except these two. I'd prefer consistency here and by that, also adjust the docs, but your choice.

File: src/renderer/fonts/font_loader.rs [hunk 3]
```diff
@@ -158,3 +165,23 @@ fn font_style(bold: bool, italic: bool) -> FontStyle {
...
+
+fn font_edging(edging: &str) -> Option<Edging> {
+    let edging = edging.to_lowercase();
+    match edging.as_str() {
+        "alias" => Some(Edging::Alias),
+        "antialias" => Some(Edging::AntiAlias),
+        "subpixelantialias" => Some(Edging::SubpixelAntiAlias),
+        _ => None
+    }
+}
```
Comment (MultisampledNight) [inline src/renderer/fonts/font_loader.rs:187]: Maybe implementing `ParseFromValue` like it's done in the cursor renderer with `VfxMode` would make this less messy and also allow including the enum _directly_ in the struct, without a potentially broken invariant.

**Review by MultisampledNight — COMMENTED** [review]
Cool, thank you! Some small things I want to be looked at before throwing this in.

**Comment by MultisampledNight** [comment]
Take yourself as much time as you need or would like to have. :heart:

File: src/renderer/mod.rs (hunk 1 above)
Comment (Nawy) [reply src/renderer/mod.rs]: Fixed formatting with fmt.  Actually, I'm running it. Don't have fmt_autosave in my nvim config. Thanks for the remark.

File: src/window/settings.rs (hunk 2 above)
Comment (Nawy) [reply src/window/settings.rs]: Done, all of them are lowercase now.

File: src/renderer/fonts/font_loader.rs (hunk 3 above)
Comment (Nawy) [reply src/renderer/fonts/font_loader.rs]: Updated with `ParseFromValue`, I don't think that it looks better. But good to have consistency.

**Comment by Nawy** [comment]
@MultisampledNight It's done. Actually, I recently discovered Neovim, and fall in love. Switched to using it for my daily work. I'm lacking a couple of things of course (including this). But going to help with that.

**Comment by MultisampledNight** [comment]
That's nice to hear, GL HF!

The `ParseFromValue` thing was _almost_ what I meant, I'll just quickly patch it myself (by putting the enum itself in the settings without `String` in between).

**Comment by Nawy** [comment]
@MultisampledNight I got it! Let me update it.

**Comment by MultisampledNight** [comment]
No need to duplicate work, I'm almost done.

**Comment by MultisampledNight** [comment]
...distractions, experiments and build times on this machine

I just found out that a `:set guifont=...` is required after setting one of those options, probably due to the font pair loading mechanisms. Maybe we can somehow re-trigger a reload of the current font pair if one of those options changes, I'll investigate that more before merging.

**Comment by Nawy** [comment]
@MultisampledNight Looks cool! I'm a little bit worried about font aliasing because you removed the option-style. So it means edging is all the time there. I'm not sure what's the default value for this parameter. 

 What do you mean by "required"? I run this build on windows/macOS with no `:set guifont` in my config. I see some double initialization of these parameters, but probably because of loading multiple fonts.

**Comment by MultisampledNight** [comment]
About "required": I wasn't precise enough, try setting `:let g:neovide_font_edging="alias"`. The change doesn't seem like it'd be applied until you run `:set guifont=whatever`.

**Comment by MultisampledNight** [comment]
...actually, now I'm wondering whether it would be better to integrate the newly introduced config variables into `guifont` directly.

**Comment by MultisampledNight** [comment]
To be honest, I think option 2 is preferrable (while removing the `edging-` part for the (anti)aliasing). The `guifont` option is kind of _meant_ to be stuffed with all font configuration, looking at the current fallback mechanism and the like. Thus, you can split it in your config on multiple lines anyways if you don't feel like reading/editing one large line.

**Comment by MultisampledNight** [comment]
Feel free to do so (and also argument for one or another should you prefer option 1), but I can do that should that be out of your current motivation fuel reach.

**Comment by Nawy** [comment]
I prefer option 1 because the font edging and hinting sound like a type of renderer. The same I would say about size, bold, and italic settings, they sound like renderer types. 
But if we prefer option 2, how would you detect FontEdging and FontHinting? I see not some literals (h, b, i) are used. So we have to use at least a similar mechanism.
I'm going to try to implement it. Give me a shot. Just would be nice to deal with the format.

**Comment by MultisampledNight** [comment]
Option 1 has the problem of these variables being seemingly unrelated to the guifont option, and we suddently have 2 places for configuring font settings.

If you want to go for option 2, this scheme sounds good to me:

- `hinting-full`, `hinting-normal`, `hinting-slight` and `hinting-none` are tokens for the appropiate hinting strategy
- `alias`, `antialias`, `subpixel`/`subpixelantialias` are tokens for the appropiate antialiasing strategy

If you want to include "edging" for the aliasing either way, fair enough. But I'd say option 2 uses existing machinery and makes it easy for users to set one, comprehensive guifont config, while option 1 adds partly split machinery where one part relies on the other one to be applied.

**Comment by Nawy** [comment]
@MultisampledNight Okay, I integrated edging and hinting into `guifont`. It works at runtime. Also, a few tests have been added to cover `FontOptions`.

**Review by MultisampledNight — COMMENTED** [review]
Almost perfect! Small things left, but otherwise I think we can get this going. Format seemed a bit odd at first due to the "special-seeming" `#` prefix, but I realized that's only for avoiding ambiguity with `hXX`.

**Comment by MultisampledNight** [comment]
Fwiw, the build issued a few warnings about a few unused imports as well:

```
warning: unused import: `error`
 --> src/renderer/fonts/font_loader.rs:3:11
  |
3 | use log::{error, trace};
  |           ^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

warning: unused import: `nvim_rs::Value`
 --> src/renderer/fonts/font_loader.rs:5:5
  |
5 | use nvim_rs::Value;
  |     ^^^^^^^^^^^^^^

warning: `neovide` (bin "neovide") generated 2 warnings
```

**Comment by MultisampledNight** [comment]
(for a few minutes I was confused on whether the changes had actually any effect when using them, until I realized I'm running a HiDPI display so the changes were barely visible, heh)

File: website/docs/configuration.md [hunk 4]
```diff
@@ -42,39 +42,20 @@ as such it's also documented in `:h guifont`. But to sum it up and also add Neov
     - `hX` — Sets the font size to `X` points, while `X` can be any (even floating-point) number.
     - `b` — Sets the font **bold**.
     - `i` — Sets the font _italic_.
+    - `#e-X` — Sets edge pixels draw opaque or with partial transparency, while `X` is a type of edging:
```
[the remark this answers is not in the document]
Comment (Nawy) [reply website/docs/configuration.md]: True - updated

File: src/renderer/fonts/font_loader.rs [hunk 5]
```diff
@@ -51,11 +46,36 @@ impl PartialEq for FontPair {
...
+        edging: FontEdging,
+    ) -> Self {
+        Self {
+            bold,
+            italic,
+            family_name,
+            hinting,
+            edging,
+        }
+    }
```
[the remark this answers is not in the document]
Comment (Nawy) [reply src/renderer/fonts/font_loader.rs]: Updated. I kinda agree with your point. Just wanted to help with these comments. But if it's just a data struct - fine.

**Comment by Nawy** [comment]
image

**Comment by Nawy** [comment]
Fixed the line size.

**Comment by Nawy** [comment]
Awesome! Now it's ready to merge. Thanks for your support @MultisampledNight

**Comment by MultisampledNight** [comment]
Chooooooo Chooooooo! Thank you for the patch!

**Comment by oredaze** [comment]
I have tried every combination with every font setting and it does not seem to remove the bold-like font rendering.

**Pull request merged**